exchanges/bitmex/bitmex_websocket.go

```diff
@@ -131,338 +152,321 @@ func (b *Bitmex) wsReadData() {
+		}
+		return nil
 	}
 
-	var respError WebsocketErrorResponse
-	if _, ok := quickCapture["status"]; ok {
-		err = json.Unmarshal(respRaw, &respError)
+	tableName, err := jsonparser.GetString(msg, "table")
+	if err != nil {
+		// Anything that's not a table isn't expected
+		return fmt.Errorf("unknown message format: %s", msg)
+	}
+
+	switch tableName {
+	case bitmexWSOrderbookL2, bitmexWSOrderbookL225, bitmexWSOrderbookL10:
+		var orderbooks OrderBookData
+		err = json.Unmarshal(msg, &orderbooks)
 		if err != nil {
 			return err
 		}
-	}
+		if len(orderbooks.Data) == 0 {
+			return fmt.Errorf("empty orderbook data received: %s", msg)
+		}
 
-	if _, ok := quickCapture["success"]; ok {
-		var decodedResp WebsocketSubscribeResp
-		err = json.Unmarshal(respRaw, &decodedResp)
+		var pair currency.Pair
+		var a asset.Item
+		pair, a, err = b.GetPairAndAssetTypeRequestFormatted(orderbooks.Data[0].Symbol)
 		if err != nil {
 			return err
 		}
 
-		if decodedResp.Success {
-			if len(quickCapture) == 3 {
-				if b.Verbose {
-					log.Debugf(log.ExchangeSys, "%s websocket: Successfully subscribed to %s",
-						b.Name, decodedResp.Subscribe)
-				}
-			} else {
-				b.Websocket.SetCanUseAuthenticatedEndpoints(true)
-				if b.Verbose {
-					log.Debugf(log.ExchangeSys, "%s websocket: Successfully authenticated websocket connection",
-						b.Name)
-				}
-			}
+		err = b.processOrderbook(orderbooks.Data, orderbooks.Action, pair, a)
+		if err != nil {
+			return err
+		}
+	case bitmexWSTrade:
+		if !b.IsSaveTradeDataEnabled() {
 			return nil
 		}
-
-		b.Websocket.DataHandler <- fmt.Errorf("%s websocket error: Unable to subscribe %s",
-			b.Name, decodedResp.Subscribe)
-	} else if _, ok := quickCapture["table"]; ok {
-		var decodedResp WebsocketMainResponse
-		err = json.Unmarshal(respRaw, &decodedResp)
+		var tradeHolder TradeData
+		err = json.Unmarshal(msg, &tradeHolder)
 		if err != nil {
 			return err
 		}
-		switch decodedResp.Table {
-		case bitmexWSOrderbookL2, bitmexWSOrderbookL225, bitmexWSOrderbookL10:
-			var orderbooks OrderBookData
-			err = json.Unmarshal(respRaw, &orderbooks)
+		var trades []trade.Data
+		for i := range tradeHolder.Data {
+			if tradeHolder.Data[i].Price == 0 {
+				// Please note that indices (symbols starting with .) post trades at intervals to the trade feed.
+				// These have a size of 0 and are used only to indicate a changing price.
+				continue
+			}
+			var p currency.Pair
+			var a asset.Item
+			p, a, err = b.GetPairAndAssetTypeRequestFormatted(tradeHolder.Data[i].Symbol)
 			if err != nil {
 				return err
 			}
-			if len(orderbooks.Data) == 0 {
-				return fmt.Errorf("%s - Empty orderbook data received: %s", b.Name, respRaw)
+			var oSide order.Side
+			oSide, err = order.StringToOrderSide(tradeHolder.Data[i].Side)
+			if err != nil {
+				b.Websocket.DataHandler <- order.ClassificationError{
+					Exchange: b.Name,
+					Err:      err,
+				}
```

Comment:
```suggestion
				return err
```

Usually only for actual orders matched with your account.

Reply:
Fixed 3377e351